includes/class-omise-callback.php

```diff
@@ -0,0 +1,178 @@
+<?php
+
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * @since 4.0
+ */
+class Omise_Callback {
+	/**
+	 * @var \Omise_Callback
+	 */
+	protected static $the_instance = null;
+
+	/**
+	 * @static
+	 *
+	 * @return \Omise_Callback - The instance.
+	 */
+	public static function instance() {
+		if ( is_null( self::$the_instance ) ) {
+			self::$the_instance = new self();
+		}
+
+		return self::$the_instance;
+	}
+
+	public static function execute() {
+		$callback = self::instance();
+
+		$order   = $callback->retrieve_order();
+		$payment = $callback->retrieve_payment( $order );
+
+		$order->add_order_note( __( 'OMISE: Validating the payment result...', 'omise' ) );
+
+		try {
+			$charge = OmiseCharge::retrieve( $order->get_transaction_id() );
+
+			$resolving_method = strtolower( 'payment_' . $charge['status'] );
```

Comment:
Wouldn't it be better to use a switch case (thus creating an explicit whitelist) here instead of allowing "untrusted" strings to call methods

Reply:
Hmm, it seems to me the same.
Using switch cases we will have to define an exact condition for its case anyway, which, relying on the same "untrusted" string to meet its condition.

With this, those that extend this class (in the future) can simply add `payment_expired` method to handle those expired charges without a need to modify this class directly.

Unless, there is something relating to security concern that I have missed it 🤔

Reply:
It might not be too bad at the moment, but you never know how the code will evolve in the future. And you won't be thinking about this bit of dynamic behavior which could lead to a catastrophic scenario.

The advice we have in other repos is that if the string is untrusted (basically it's not coming from the repo itself and instead is a string coming from the outside) don't use it to call methods dynamically.

The most likely person to implement that `expired` scenario is you. So I wouldn't worry too much about extensibility :troll:

The other reason I don't like using this is that then you're using an exception to do flow control instead of using the else branch of the switch to define what is unknown behavior.

Reply:
@robinclart actually the reason we need `try ... catch` there is because `OmiseCharge::retrieve()` will throw an exception in case if it cannot retrieve a charge object from Omise API, or else we wouldn't be able to catch an error message from Omise API, `error` object. Without wrapping everything under `exception` flow, then the code could have turn into some mess like this
```php
try {
	$charge = OmiseCharge::retrieve( $order->get_transaction_id() );
} catch ( Exception $e ) {
	$order->add_order_note(
		sprintf(
			wp_kses( __( 'OMISE: Unable to validate the result.<br/>%s', 'omise' ), array( 'br' => array() ) ),
			$e->getMessage()
		)
	);

	$callback->invalid_result();
}

switch ($charge['status']) {
	case 'successful':
	case 'failed':
	case 'pending':
		$resolving_method = strtolower( 'payment_' . $charge['status'] );
		$callback->$resolving_method( $order, $charge );
		break;
	
	default:
		$callback->invalid_result();
		break;
}
```
 Which, it feels a bit irritating to see this mess... but, the point that you mentioned about using exception as flow control is a bad idea is interesting. Would you mind to explain about that a bit? (I've seen some debate over internet, but just wonder how do you view about it)

Reply:
Right. Read the code too fast. Strike my last point then.